src/instruction.rs

```diff
@@ -0,0 +1,268 @@
+//! Program instruction types.
+
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+/// Instructions supported by the Paladin Rewards program.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PaladinRewardsInstruction {
+    /// Configures rewards for a mint that has been configured with the rewards
+    /// program as a transfer hook program.
+    ///
+    /// This instruction will:
+    ///
+    /// - Initialize a rewards account for the mint (distribution account),
+    ///   configured with the distribution addresses:
+    ///   - Active rewards
+    ///   - Piggy bank
+    ///   - Staked PAL rewards
+    /// - Initialize an active rewards account for the mint.
+    /// - Initialize the required accounts for the transfer hook.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Distribution account.
+    /// 1. `[w]` Active rewards account.
+    /// 2. `[w]` Transfer hook extra account metas account.
+    /// 3. `[ ]` Token mint.
+    /// 4. `[s]` Mint authority.
+    InitializeMintRewardInfo {
+        piggy_bank_address: Pubkey,
+        staked_rewards_address: Pubkey,
+    },
+    /// Moves the active rewards to the distribution account and updates total
+    /// rewards.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[ ]` Token mint.
+    UpdateMintRewards,
+    /// Moves SOL rewards to the following parties:
+    ///
+    /// - 1%  Piggy bank
+    /// - 4%  Staked PAL (validators)
+    /// - 5%  All PAL holders
+    /// - 90% Leader who produces the block
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[w]` Piggy bank account.
+    /// 3. `[w]` Staked PAL rewards account.
+    /// 4. `[w]` Leader account.
```

Comment:
Ah darn, you know what, seeing all these write locks, I realize that it doesn't make a ton of sense to separate the active rewards account, since you need a write lock on the distribution account here regardless.

To avoid all these write locks, the alternative is to only take writelocks on the active rewards account and the leader account, and then use `UpdateMintRewards` to distribute to the other addresses. What do you think?

Reply:
Doesn't that just shift all the write locks to `UpdateMintRewards` except for one? The leader account? Would we also be removing the concept of an active rewards account then?

Reply:
Correct, but the idea is that `UpdateMintRewards` won't be called as often as this instruction. I'd say there's two options:

* remove the active rewards account, distribute everything here
* include the active rewards account, no other write locks here

I'd lean towards the second option, but not by a lot

Reply:
As @edgarpavlovsky [mentioned below](https://github.com/paladin-bladesmith/paladin-rewards-program/pull/1#issuecomment-2138200693), we've agreed to cut the active rewards account and the concept of sweeping.